File: apps/api/app/services/source_filters.py
from dataclasses import dataclass


CANONICAL_SOURCE_TYPES = (
    "byu_speeches_es",
    "byu_speeches_en",
    "discursos_sud",
    "general_conference",
    "church_manuals",
    "joseph_smith_papers",
    "byu_rsc",
)


SOURCE_ALIASES = {
    "byu_speeches_es": {"byu_speeches_es", "byu_speeches_spanish", "speeches_byu_es"},
    "byu_speeches_en": {"byu_speeches_en", "byu_speeches", "speeches_byu", "byu"},
    "discursos_sud": {"discursos_sud", "discurso_sud", "discursosud"},
    "general_conference": {"general_conference", "conference", "conferencia_general", "general-conference"},
    "church_manuals": {"church_manuals", "manuals", "manuales", "church_manual"},
    "joseph_smith_papers": {"joseph_smith_papers", "jsp", "josephsmithpapers"},
    "byu_rsc": {"byu_rsc", "rsc", "religious_studies_center"},
}

SOURCE_LABELS = {
    "byu_speeches_es": "BYU Speeches ES",
    "byu_speeches_en": "BYU Speeches EN",
    "discursos_sud": "Discursos SUD",
    "general_conference": "Conferencia General",
    "church_manuals": "Manuales de la Iglesia",
    "joseph_smith_papers": "Joseph Smith Papers",
    "byu_rsc": "BYU Religious Studies Center",
}


@dataclass(frozen=True)
class SourceOption:
    key: str
    label: str
    aliases: tuple[str, ...]


def normalize_source_type(value: str | None) -> str | None:
    if not value:
        return None
    normalized = value.strip().lower().replace("-", "_").replace(" ", "_")
    for canonical, aliases in SOURCE_ALIASES.items():
        if normalized == canonical or normalized in aliases:
            return canonical
    return normalized


def source_type_aliases(value: str | None) -> list[str]:
    normalized = normalize_source_type(value)
    if not normalized:
        return []
    aliases = set(SOURCE_ALIASES.get(normalized, set()))
    aliases.add(normalized)
    return sorted(aliases)


def canonical_source_options() -> list[SourceOption]:
    return [
        SourceOption(
            key=key,
            label=SOURCE_LABELS[key],
            aliases=tuple(sorted(SOURCE_ALIASES.get(key, {key}))),
        )
        for key in CANONICAL_SOURCE_TYPES
    ]

File: apps/web/lib/source-filters.ts
export type SourceFilterOption = {
  key: string;
  label: string;
  documentCount?: number;
  canonical?: boolean;
  aliases?: string[];
};

export const canonicalSourceOptions: SourceFilterOption[] = [
  { key: "byu_speeches_es", label: "BYU Speeches ES" },
  { key: "byu_speeches_en", label: "BYU Speeches EN" },
  { key: "discursos_sud", label: "Discursos SUD" },
  { key: "general_conference", label: "Conferencia General" },
  { key: "church_manuals", label: "Manuales de la Iglesia" },
  { key: "joseph_smith_papers", label: "Joseph Smith Papers" },
  { key: "byu_rsc", label: "BYU Religious Studies Center" }
];

export function mergeSourceOptions(items?: SourceFilterOption[]): SourceFilterOption[] {
  if (!items?.length) {
    return canonicalSourceOptions;
  }
  const byKey = new Map(canonicalSourceOptions.map((item) => [item.key, item]));
  for (const item of items) {
    byKey.set(item.key, { ...byKey.get(item.key), ...item });
  }
  return Array.from(byKey.values()).sort((a, b) => a.label.localeCompare(b.label));
}
